node/src/components/consensus/highway_core/active_validator.rs

```diff
@@ -478,82 +483,268 @@ mod tests {
         }
     }
 
-    fn unwrap_single<T: Debug>(vec: Vec<T>) -> T {
-        let mut iter = vec.into_iter();
+    struct TestState {
+        rng: TestRng,
+        state: State<TestContext>,
+        instance_id: u64,
+        fd: FinalityDetector<TestContext>,
+        validators: HashMap<ValidatorIndex, ActiveValidator<TestContext>>,
+        timers: BTreeSet<(Timestamp, ValidatorIndex)>,
+    }
+
+    impl TestState {
+        fn new(
+            state: State<TestContext>,
+            rng: TestRng,
+            start_time: Timestamp,
+            instance_id: u64,
+            fd: FinalityDetector<TestContext>,
+            validators: Vec<ValidatorIndex>,
+        ) -> Self {
+            let mut timers = BTreeSet::new();
+            let earliest_round_start = state::round_id(start_time, state.params().init_round_exp())
+                + (1 << state.params().init_round_exp()).into();
+            let validator_map = validators
+                .into_iter()
+                .map(|vidx| {
+                    let secret = TestSecret(vidx.0);
+                    let (av, effects) = ActiveValidator::new(vidx, secret, start_time, &state);
+                    if let [Eff::ScheduleTimer(timestamp)] = *effects {
+                        // We start at time 410, with round length 16, so the first leader tick is
+                        // 416, and the first witness tick 426.
+                        if state.leader(earliest_round_start) == vidx {
+                            assert_eq!(
+                                timestamp,
+                                416.into(),
+                                "Invalid initial timer scheduled for {:?}.",
+                                vidx
+                            )
+                        } else {
+                            let witness_offset = av
+                                .witness_offset(state::round_len(state.params().init_round_exp()));
+                            let witness_timestamp = earliest_round_start + witness_offset;
+                            assert_eq!(
+                                timestamp, witness_timestamp,
+                                "Invalid initial timer scheduled for {:?}.",
+                                vidx
+                            )
+                        }
+                        timers.insert((timestamp, vidx));
+                    } else {
+                        panic!("Only one initial effect expected")
```

Comment:
Panicking earlier would make it easier to remember which check failed (and reduce the indentation).